File: 16-3sum-closest/3sum-closest.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        n = len(nums)
        diff = float('inf')
        res_sum = 0

        for i in range(n - 2):
            left = i + 1
            right = n - 1

            while left < right:
                total = nums[i] + nums[left] + nums[right]
                d = abs(target - total)

                # Update the closest sum if a smaller difference is found
                if diff > d:
                    diff = d
                    res_sum = total

                # Exact match found
                if total == target:
                    return res_sum

                if total < target:
                    left += 1
                else:
                    right -= 1

        return res_sum

        

File: 16-3sum-closest/3sum-closest.cpp
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        sort(nums.begin(),nums.end());
        int answer = nums[0]+nums[1]+nums[2];
        for(int i =0;i<nums.size()-2;i++){
            int j = i+1,k=nums.size()-1;
            while(j<k){
                int sum = nums[i]+nums[j]+nums[k];
                if(abs(target-sum)<abs(target-answer))
                answer = sum;
                if(sum==target) return target;
                else if(sum<target) j++;
                else k--;
            }
        }
        return answer;
    }
};